README.md
# PLHW Api Client

PLHW API Client provides means to authenticate clients against our OAuth2 server and issue authorized requests to our api endpoints.


## Installation

```bash
composer require plhw/hf-api-client
```

## Example API calls

Once you have succesfully installed the application, some example scripts are located at. To use them first copy an configuration file to your app root. 

```bash
cp ./vendor/plhw/hf-api-client/example/.hf-api-client-secrets.php.dist ./.hf-api-client-secrets.php
```

Open `.hf-api-client-secrets.php` and configure it with credentials you got from us.

Finally the example scripts are configured to use `data/cache` as directory for its cached accss tokens. This directory must exist (only for the examples).

Now you can run the example scripts;

```bash
php ./vendor/plhw/hf-api-client/example/practicesPos.php
```

## Usage

To implement usage in your application you can have a look at the example scripts.

```php
$options = \HF\ApiClient\Options\Options::fromArray(
    [
        'server_uri'    => 'https://api.plhw.nl',
        'client_id'     => 'id',
        'client_secret' => 'secret',    
        'scope'         => 'customer_pos',
        'grant_type'    => 'client_credentials',
    ]
);

$api = \HF\ApiClient\ApiClient::createClient($options, /* $cache OPTIONAL */);
```

The above is all that is nessesary to configure, though I recommend you tweak the caching meganism. Currently we use the [zendframework/zend-cache](https://docs.zendframework.com/zend-cache/) component. This might change when v3 is released to use the psr-7 caching FIG standard.

When you have `ApiClient` instance you can use it by calling methods as defined in `/vendor/plhw/hf-api-client/data/v1` on it. Additionally [API documentation](https://api.plhw.nl/docs) can be found here.

for example, search the published practices around a coordinate.

```php
$results = $api->customer_posAroundCoordinate('52.3629882,4.8593175', 15000, 'insoles');

if ($api->isSuccess()) {
    foreach ($results['data'] as $result) {
        printf("Practice %s on %skm\n", $result['attributes']['name'],
            round(($result['attributes']['distance'] / 100)) / 10);
        printf(" - sells %s\n", implode(', ', $result['attributes']['products']));
    }
} else {
    printf("Error (%d): %s\n", $api->getStatusCode(), $result);
}
```

## Under the hood

When you call any method on the APIClient instance an access token is requested from our OAuth2 server. This access token is then cached for aditional uses up to the moment it expires or is deleted. It will then get a new access token.

Any calls to our API are now `signed` with that access token and is used by our permission system to determain if you have access or not.

Our API will accept and return json payload with are automaticly (de)encoded.

### OAuth2 ClientCredentialsGrant

For machine to machine communication OAuth2 ClientCredentialGrant is appropiate. 
You must obtain the following information from us.

1. client ID
2. client secret
3. scope

Your client side application should communicate via a proxy to our server.

```

               
   [webbrowser] <----- internet ----->   [application]   <----- internet -----> [api.plhw.nl]

1.  request data    -> request ->      do we have valid access token?

2.                                     NO, get access token   -> request  -> id,secret,scope

3.                                                                                validate request

4.                                     store token in cache   <- response <- access token
                                                                     
5.                                     YES, request & token.  -> request  -> validate token

6.                                     data                   <- response <- data

7.  data             <- response <-    data

```




Run all examples in the 'commerce' domain:

```
find ./example/commerce/ -maxdepth 1 -type f -exec {} \;
```

## Upgrading

v2 has been rewritten for php 8+. Principals are the same but there are a few minor changes that will break your application.

Please see our examples to see how you interact with the api.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.4.0](https://github.com/plhw/hf-api-client/compare/v2.3.1...v2.4.0) (2021-06-29)


### Features

* added getOrderLead and queryOrderLeads ([e13ea9f](https://github.com/plhw/hf-api-client/commits/e13ea9fcb6e157383a64c13326717581f20d4e00))

### [2.3.1](https://github.com/plhw/hf-api-client/compare/v2.3.0...v2.3.1) (2021-06-28)

## [2.3.0](https://github.com/plhw/hf-api-client/compare/v2.2.0...v2.3.0) (2021-06-24)


### Features

* ability to change more customer and practice properties ([116f21d](https://github.com/plhw/hf-api-client/commits/116f21db9465bf3f3b7470a3cc03d0ed49238021))

## [2.2.0](https://github.com/plhw/hf-api-client/compare/v2.1.0...v2.2.0) (2021-06-22)


### Features

* ability to modify addresses ([d664aee](https://github.com/plhw/hf-api-client/commits/d664aee2bde3bd60530da5478a0c3030c832fe6c))

## [2.1.0](https://github.com/plhw/hf-api-client/compare/v2.0.1...v2.1.0) (2021-06-08)


### Features

* show example of order-lead submition ([759861a](https://github.com/plhw/hf-api-client/commits/759861a2a92ac2bccc1896e5fac51d741da05daf))


### Bug Fixes

* token invalidation ([23b1a46](https://github.com/plhw/hf-api-client/commits/23b1a46b97bff00038bbda0f78a3b7673353fffc))

### 2.0.1 (2021-06-08)


example/practicesPos.php
#!/usr/bin/env php
<?php

declare(strict_types=1);

$autoloadFiles = [
    __DIR__ . '/../../../../vendor/autoload.php',
    __DIR__ . '/../../../vendor/autoload.php',
    __DIR__ . '/../../vendor/autoload.php',
    __DIR__ . '/../vendor/autoload.php',
    __DIR__ . '/vendor/autoload.php',
];

foreach ($autoloadFiles as $autoloadFile) {
    if (file_exists($autoloadFile)) {
        chdir(dirname($autoloadFile) . '/..');
        require_once $autoloadFile;
    }
}

use HF\ApiClient\ApiClient;
use HF\ApiClient\Options\Options;
use Zend\Cache\StorageFactory;

if (! file_exists('.hf-api-client-secrets.php')) {
    die('copy example/.hf-api-client-secrets.php.dist to APP_ROOT/.hf-api-client-secrets.php');
}

// optional but will then use filesystem default tmp directory
$cache = StorageFactory::factory([
    'adapter' => [
        'name'    => 'filesystem',
        'options' => [
            'cache_dir' => './data/cache',
            'dir_level' => 0,
            'namespace' => 'a',
        ],
    ],
    'plugins' => ['serializer'],
]);

$options = Options::fromArray(include('.hf-api-client-secrets.php'));

$api = ApiClient::createClient($options, $cache);

$query = \HF\ApiClient\Query\Query::create()
    ->withFilter('around', '52.3629882,4.8593175')
    ->withFilter('distance', 50000)
    ->withFilter('product', 'insoles')
    ->withPage(1, 3)
    ->withSort('name', false);

try {
    $results = $api->customer_posAroundCoordinate($query);
} catch (\League\OAuth2\Client\Provider\Exception\IdentityProviderException $e) {
    die($e->getMessage());
} catch (\HF\ApiClient\Exception\GatewayException $e) {
    printf("%s\n\n", $e->getMessage());
    printf('%s', $api->getLastResponseBody());
    die();
}

if ($api->isSuccess()) {
    foreach ($results['data'] as $result) {
        printf("Practice %s on %skm\n", $result['attributes']['name'],
            round(($result['attributes']['distance'] / 100)) / 10);
        printf(" - sells %s\n", implode(', ', $result['attributes']['products']));
    }
} else {
    printf("Error (%d)\n", $api->getStatusCode());
    print_r($results);
}
